llarp/messages/relay_status.cpp

```diff
@@ -0,0 +1,295 @@
+#include <messages/relay_status.hpp>
+
+#include <crypto/crypto.hpp>
+#include <messages/path_confirm.hpp>
+#include <path/path.hpp>
+#include <router/abstractrouter.hpp>
+#include <util/bencode.hpp>
+#include <util/buffer.hpp>
+#include <util/logger.hpp>
+#include <util/logic.hpp>
+#include <nodedb.hpp>
+
+#include <functional>
+
+namespace llarp
+{
+  struct LRSM_AsyncHandler : public std::enable_shared_from_this< LRSM_AsyncHandler >
+  {
+    using HopHandler_ptr = std::shared_ptr< llarp::path::IHopHandler >;
+
+    std::array< EncryptedFrame, 8 > frames;
+    PathID_t pathid;
+    uint64_t status = 0;
+    HopHandler_ptr path;
+    AbstractRouter *router;
+
+    LRSM_AsyncHandler(const std::array< EncryptedFrame, 8> &_frames,
+                      const PathID_t& _pathid, uint64_t _status,
+                      HopHandler_ptr _path, AbstractRouter *_router)
+        : frames(_frames), pathid(_pathid), status(_status), path(_path), router(_router)
+    {
+    }
+
+    ~LRSM_AsyncHandler()
+    {
+    }
+
+    void handle()
+    {
+      path->HandleLRSM(status, frames, router);
+    }
+
+    void queue_handle()
+    {
+      auto func = std::bind(&llarp::LRSM_AsyncHandler::handle, shared_from_this());
+      llarp_threadpool_queue_job(router->pathContext().Worker(), func);
+    }
+
+  };
+
+
+  LR_StatusMessage::~LR_StatusMessage()
+  {
+  }
+
+  bool
+  LR_StatusMessage::DecodeKey(const llarp_buffer_t& key, llarp_buffer_t* buf)
+  {
+    bool read = false;
+    if(key == "c")
+    {
+      return BEncodeReadArray(frames, buf);
+    }
+    else if(key == "p")
+    {
+      if (!BEncodeMaybeReadDictEntry("p", pathid, read, key, buf))
+      {
+        return false;
+      }
+    }
+    else if(key == "s")
+    {
+      if (!BEncodeMaybeReadDictInt("s", status, read, key, buf))
+      {
+        return false;
+      }
+    }
+    else if(key == "v")
+    {
+      if(!BEncodeMaybeReadVersion("v", version, LLARP_PROTO_VERSION, read, key,
+                                  buf))
+      {
+        return false;
+      }
+    }
+
+    return read;
+  }
+
+  void
+  LR_StatusMessage::Clear()
+  {
```

Comment:
```c++
std::for_each(frames.begin(), frames.end(), & EncryptedFrame::Clear);
```

Reply:
As above, copied from LR_CommitMessage, should I change both?

Reply:
i think they might be fixed in #650

Reply:
addressed